website/content/users/qsim.md

```diff
@@ -0,0 +1,592 @@
+
+```cpp
+bool VqeWorkflow::initialize(const HeterogeneousMap &params) {
+  const std::string DEFAULT_OPTIMIZER = "nlopt";
+  optimizer.reset();
+  if (params.pointerLikeExists<Optimizer>("optimizer")) {
+    optimizer =
+        xacc::as_shared_ptr(params.getPointerLike<Optimizer>("optimizer"));
+  } else {
+    optimizer = createOptimizer(DEFAULT_OPTIMIZER);
+  }
+  config_params = params;
+  // VQE workflow requires an optimizer
+  return (optimizer != nullptr);
+}
+
+QuantumSimulationResult
+VqeWorkflow::execute(const QuantumSimulationModel &model) {
+  auto nParams = model.user_defined_ansatz->nParams();
+  evaluator = getEvaluator(model.observable, config_params);
+
+  OptFunction f(
+      [&](const std::vector<double> &x, std::vector<double> &dx) {
+        auto kernel = model.user_defined_ansatz->evaluate_kernel(x);
+        auto energy = evaluator->evaluate(kernel);
+        return energy;
+      },
+      nParams);
+
+  auto result = optimizer->optimize(f);
+  return {{"energy", result.first}, {"opt-params", result.second}};
+}
+```
+
+As previously described, the VQE algorithm requires a classical optimizer hence we need to figure out which optimizer should be used during initialization.  In this case, we look for the `optimizer` key if provided. Otherwise, we just fall back to a default one (`nlopt` in this case).
+
+Similarly, one may preset (providing default values) and parse any number of configuration parameters that his custom workflow support.
+
+During workflow's `execute`, one can take advantage of QCOR API's to construct the quantum circuit. In this simple VQE workflow, the ansatz circuit was provided as a QCOR kernel functor (`user_defined_ansatz`), hence, we just need to evaluate (resolving variational gate parameters) during the optimization loop.
+
+One key element of implementing workflow execution is to hook up the appropriate cost function evaluator. In that regard, QSim provides a utility function `getEvaluator` which will pick the appropriate evaluator based on user configurations, e.g., selecting the `default` (tomography-based) evaluator if none provided. 
+
+Workflow developers are free to *not* use the `getEvaluator` utility if the workflow doesn't need to evaluate (observe) the operator or requires custom logic in selecting the evaluator.
+
+Workflow developers are free to select which information to be returned at the end of the workflow execution. In particular, the workflow returns a heterogeneous key-value map (dictionary). In the VQE example, we returned the optimized energy value as well as the optimal parameters.
+
+Lastly, one needs to register the new workflow implementation with the QSim library so that users can retrieve the new workflow via the QSim workflow registry (via QSim `getWorkflow` function).
+
+This can be achieved by adding a `RegisterService` entry to the `QuantumSimulationActivator::Start` method, e.g.
```

Comment:
Can we convert the following to use `REGISTER_PLUGIN(qsim::VqeWorkflow, qsim::QuantumSimulationWorkflow)`